go-controller/pkg/libovsdbops/model_client.go

```diff
@@ -47,43 +47,77 @@ func extractUUIDsFromModels(models interface{}) []string {
 }
 
 // buildMutationsFromFields builds mutations that use the fields as values.
-func buildMutationsFromFields(fields []interface{}, mutator ovsdb.Mutator) []model.Mutation {
-	mutations := make([]model.Mutation, 0, len(fields))
+func buildMutationsFromFields(fields []interface{}, mutator ovsdb.Mutator) ([]model.Mutation, error) {
+	mutations := []model.Mutation{}
 	for _, field := range fields {
-		v := reflect.ValueOf(field)
-		if v.Kind() != reflect.Ptr {
-			panic(fmt.Sprintf("Expected Ptr but got %s", v.Kind()))
-		}
-		if v.IsNil() || v.Elem().IsNil() {
-			continue
-		}
-
-		if m, ok := field.(*map[string]string); ok {
-			// check if all values on m are zero, if so create slice of keys of m and set that to field
-			allEmpty := true
-			keySlice := []string{}
-			for key, value := range *m {
-				keySlice = append(keySlice, key)
-				if len(value) > 0 {
-					allEmpty = false
-					break
+		switch v := field.(type) {
+		case *map[string]string:
+			if v == nil || len(*v) == 0 {
+				continue
+			}
+			if mutator == ovsdb.MutateOperationDelete {
+				// turn empty map values into a mutation to remove the key for
+				// delete mutations
+				removeKeys := make([]string, 0, len(*v))
+				updateKeys := make(map[string]string, len(*v))
+				for key, value := range *v {
+					if value == "" {
+						removeKeys = append(removeKeys, key)
+					} else {
+						updateKeys[key] = value
+					}
 				}
+				if len(removeKeys) > 0 {
+					mutation := model.Mutation{
+						Field:   field,
+						Mutator: mutator,
+						Value:   removeKeys,
+					}
+					mutations = append(mutations, mutation)
```

Comment:
shouldn't append this mutation only when `updateKeys` is empty ?

Reply:
No, we are splitting the updates into two mutations: one for key/values to be added and another one for keys to be removed. We were not doing this before but no reason not to have it. Check the unit test case:
https://github.com/ovn-org/ovn-kubernetes/blob/205bb968eebd40a338d1a86bb72e3c9ba41069a3/go-controller/pkg/libovsdbops/model_client_test.go#L1388

Reply:
@jcaamano thanks for the explanation. so it's going to remove map entries in both cases as field is a map, am i right ? 
why can't just provide slice of keys in both cases ?

Reply:
Yes.
Because sometimes we want to remove keys irrespective of the value. Other times we only want to remove when both key and value match.
From OVSDB RFC:
```
         If <mutator> is "delete", then <value> may have the same type
         as <column> (a map type), or it may be a set whose element type
         is the same as <column>'s key type:

         +  If <value> is a map, the mutation deletes each key-value
            pair in <column> whose key and value equal one of the key-
            value pairs in <value>.

         +  If <value> is a set, the mutation deletes each key-value
            pair in <column> whose key equals one of the values in
            <value>.
```

Reply:
ok @jcaamano , it explains. thank you.